File: electron/vault/notebooks.ts
import fs from 'node:fs';
import path from 'node:path';
import { resolveVaultPath } from './index';
import { softDelete } from './trash';
import { readIgnorePatterns, addIgnorePattern, isIgnored } from './ignore';

export interface NotebookInfo {
  name: string;
  path: string;
  stack: string | null;
  noteCount: number;
}

export { addIgnorePattern };

const EXCLUDED_DIRS = new Set(['.thoughtstack', '.trash', '.images']);

/**
 * Creates a notebook directory. If stack is provided, creates inside the stack directory.
 */
export function createNotebook(vaultPath: string, name: string, stack?: string): NotebookInfo {
  const resolved = resolveVaultPath(vaultPath);

  let notebookRelPath: string;
  if (stack) {
    notebookRelPath = path.join(stack, name);
  } else {
    notebookRelPath = name;
  }

  const fullPath = path.join(resolved, notebookRelPath);
  fs.mkdirSync(fullPath, { recursive: true });

  return {
    name,
    path: notebookRelPath,
    stack: stack ?? null,
    noteCount: 0,
  };
}

/**
 * Renames a notebook directory.
 */
export function renameNotebook(
  vaultPath: string,
  oldPath: string,
  newName: string
): NotebookInfo {
  const resolved = resolveVaultPath(vaultPath);
  const oldFullPath = path.join(resolved, oldPath);
  const parentDir = path.dirname(oldFullPath);
  const newFullPath = path.join(parentDir, newName);

  fs.renameSync(oldFullPath, newFullPath);

  const newRelPath = path.relative(resolved, newFullPath);
  const stack = path.dirname(newRelPath);

  // Count .md files
  const noteCount = countMdFiles(newFullPath);

  return {
    name: newName,
    path: newRelPath,
    stack: stack === '.' ? null : stack,
    noteCount,
  };
}

/**
 * Deletes a notebook directory. Moves all notes to .trash/ first, then removes the directory.
 */
export function deleteNotebook(vaultPath: string, notebookPath: string): boolean {
  const resolved = resolveVaultPath(vaultPath);
  const fullPath = path.join(resolved, notebookPath);

  if (!fs.existsSync(fullPath)) return false;

  // Move all .md files to trash first
  const entries = fs.readdirSync(fullPath, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.isFile() && entry.name.endsWith('.md')) {
      const notePath = path.join(notebookPath, entry.name);
      softDelete(vaultPath, notePath);
    }
  }

  // Remove the directory (and any remaining non-md files like .images/)
  fs.rmSync(fullPath, { recursive: true, force: true });

  return true;
}

/**
 * Moves a notebook directory to a different stack (or to vault root if targetStack is undefined).
 */
export function moveNotebook(
  vaultPath: string,
  notebookPath: string,
  targetStack?: string
): NotebookInfo {
  const resolved = resolveVaultPath(vaultPath);
  const oldFullPath = path.join(resolved, notebookPath);
  const name = path.basename(notebookPath);

  let newRelPath: string;
  if (targetStack) {
    newRelPath = path.join(targetStack, name);
  } else {
    newRelPath = name;
  }

  const newFullPath = path.join(resolved, newRelPath);

  // Ensure target parent exists
  fs.mkdirSync(path.dirname(newFullPath), { recursive: true });

  fs.renameSync(oldFullPath, newFullPath);

  const noteCount = countMdFiles(newFullPath);

  return {
    name,
    path: newRelPath,
    stack: targetStack ?? null,
    noteCount,
  };
}

/**
 * Lists directories in the vault, excluding .thoughtstack, .trash, .images.
 * Handles both top-level notebooks and notebooks inside stacks.
 */
export function listNotebooks(vaultPath: string): NotebookInfo[] {
  const resolved = resolveVaultPath(vaultPath);
  const ignorePatterns = readIgnorePatterns(resolved);
  const results: NotebookInfo[] = [];

  const entries = fs.readdirSync(resolved, { withFileTypes: true });
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    if (EXCLUDED_DIRS.has(entry.name)) continue;
    if (entry.name.startsWith('.')) continue;
    if (isIgnored(ignorePatterns, entry.name)) continue;

    const dirPath = path.join(resolved, entry.name);

    // Check if this directory contains .md files (it's a notebook)
    // or only subdirectories (it's a stack)
    const subEntries = fs.readdirSync(dirPath, { withFileTypes: true });
    const hasMdFiles = subEntries.some((e) => e.isFile() && e.name.endsWith('.md'));
    const hasSubDirs = subEntries.some(
      (e) => e.isDirectory() && !e.name.startsWith('.')
    );

    if (hasMdFiles || !hasSubDirs) {
      results.push({
        name: entry.name,
        path: entry.name,
        stack: null,
        noteCount: countMdFiles(dirPath),
      });
    }

    // If it has subdirectories, those are notebooks in a stack
    if (hasSubDirs) {
      for (const subEntry of subEntries) {
        if (!subEntry.isDirectory()) continue;
        if (subEntry.name.startsWith('.')) continue;

        const relPath = path.join(entry.name, subEntry.name).replace(/\\/g, '/');
        if (isIgnored(ignorePatterns, relPath)) continue;

        const subDirPath = path.join(dirPath, subEntry.name);
        results.push({
          name: subEntry.name,
          path: relPath,
          stack: entry.name,
          noteCount: countMdFiles(subDirPath),
        });
      }
    }
  }

  return results;
}

function countMdFiles(directory: string): number {
  try {
    const entries = fs.readdirSync(directory);
    return entries.filter((e) => e.endsWith('.md')).length;
  } catch {
    return 0;
  }
}

File: src-tauri/src/vault/notebooks.rs
use crate::vault::{
    ignore::{is_ignored, read_ignore_patterns},
    resolve_vault_path,
    sanitize::sanitize_filename,
    trash::soft_delete_dir,
    SKIP_DIRS,
};
use serde::{Deserialize, Serialize};
use std::fs;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NotebookInfo {
    pub name: String,
    pub path: String,
    pub stack: Option<String>,
    pub note_count: usize,
}

fn count_md_files(dir: &std::path::Path) -> usize {
    walkdir::WalkDir::new(dir)
        .follow_links(false)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| {
            e.file_type().is_file()
                && e.path().extension().and_then(|x| x.to_str()) == Some("md")
                && !e
                    .file_name()
                    .to_str()
                    .map(|n| n.starts_with('.'))
                    .unwrap_or(false)
        })
        .count()
}

/// Creates a notebook directory. If `stack` is provided, creates inside the stack directory.
pub fn create_notebook(
    vault_path: &str,
    name: &str,
    stack: Option<&str>,
) -> Result<NotebookInfo, String> {
    let resolved = resolve_vault_path(vault_path);

    let notebook_rel = if let Some(s) = stack {
        format!("{}/{}", s, name)
    } else {
        name.to_string()
    };

    let full_path = resolved.join(&notebook_rel);
    fs::create_dir_all(&full_path).map_err(|e| e.to_string())?;

    Ok(NotebookInfo {
        name: name.to_string(),
        path: notebook_rel,
        stack: stack.map(|s| s.to_string()),
        note_count: 0,
    })
}

/// Renames a notebook directory.
pub fn rename_notebook(
    vault_path: &str,
    old_rel_path: &str,
    new_name: &str,
) -> Result<NotebookInfo, String> {
    let resolved = resolve_vault_path(vault_path);
    let old_full = resolved.join(old_rel_path);
    let parent = old_full.parent().ok_or("Invalid notebook path")?;
    let new_full = parent.join(sanitize_filename(new_name));

    fs::rename(&old_full, &new_full).map_err(|e| e.to_string())?;

    let new_rel = new_full
        .strip_prefix(&resolved)
        .map_err(|e| e.to_string())?
        .to_string_lossy()
        .to_string();

    let parent_rel = new_full
        .strip_prefix(&resolved)
        .ok()
        .and_then(|p| p.parent())
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_default();

    let stack = if parent_rel.is_empty() || parent_rel == "." {
        None
    } else {
        Some(parent_rel)
    };

    let note_count = count_md_files(&new_full);

    Ok(NotebookInfo {
        name: new_name.to_string(),
        path: new_rel,
        stack,
        note_count,
    })
}

/// Deletes a notebook (soft-deletes to .trash/).
pub fn delete_notebook(vault_path: &str, notebook_rel_path: &str) -> Result<bool, String> {
    soft_delete_dir(vault_path, notebook_rel_path)
}

/// Moves a notebook to a different stack (or to vault root if `target_stack` is None).
pub fn move_notebook(
    vault_path: &str,
    notebook_rel_path: &str,
    target_stack: Option<&str>,
) -> Result<NotebookInfo, String> {
    let resolved = resolve_vault_path(vault_path);
    let src = resolved.join(notebook_rel_path);

    let notebook_name = src
        .file_name()
        .and_then(|n| n.to_str())
        .ok_or("Invalid notebook path")?
        .to_string();

    let dest_parent = if let Some(stack) = target_stack {
        let stack_dir = resolved.join(stack);
        fs::create_dir_all(&stack_dir).map_err(|e| e.to_string())?;
        stack_dir
    } else {
        resolved.clone()
    };

    let dest = dest_parent.join(&notebook_name);
    fs::rename(&src, &dest).map_err(|e| e.to_string())?;

    let new_rel = dest
        .strip_prefix(&resolved)
        .map_err(|e| e.to_string())?
        .to_string_lossy()
        .to_string();

    let note_count = count_md_files(&dest);

    Ok(NotebookInfo {
        name: notebook_name,
        path: new_rel,
        stack: target_stack.map(|s| s.to_string()),
        note_count,
    })
}

/// Lists all notebooks in the vault.
/// A "notebook" is any directory one level below vault root (or two levels deep
/// for stacked notebooks). Directories in SKIP_DIRS are excluded.
pub fn list_notebooks(vault_path: &str) -> Result<Vec<NotebookInfo>, String> {
    let resolved = resolve_vault_path(vault_path);
    let ignore_patterns = read_ignore_patterns(&resolved);
    let mut notebooks = Vec::new();

    let entries = fs::read_dir(&resolved).map_err(|e| e.to_string())?;

    for entry in entries.filter_map(|e| e.ok()) {
        let path = entry.path();
        if !path.is_dir() {
            continue;
        }
        let name = path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
            .to_string();

        if SKIP_DIRS.contains(&name.as_str()) || name.starts_with('.') {
            continue;
        }
        if is_ignored(&ignore_patterns, &name) {
            continue;
        }

        // Check if this looks like a stack (contains sub-directories with notes)
        // vs a direct notebook.
        let sub_dirs: Vec<_> = fs::read_dir(&path)
            .map(|rd| {
                rd.filter_map(|e| e.ok())
                    .filter(|e| e.path().is_dir() && !e.file_name().to_str().unwrap_or("").starts_with('.'))
                    .collect()
            })
            .unwrap_or_default();

        let direct_notes = count_md_files(&path);

        if direct_notes > 0 || sub_dirs.is_empty() {
            // Treat as a direct notebook
            notebooks.push(NotebookInfo {
                name: name.clone(),
                path: name.clone(),
                stack: None,
                note_count: direct_notes,
            });
        }

        // Also list any stacked sub-notebooks
        for sub_entry in sub_dirs {
            let sub_path = sub_entry.path();
            let sub_name = sub_path
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("")
                .to_string();
            let sub_rel = format!("{}/{}", name, sub_name);
            if is_ignored(&ignore_patterns, &sub_rel) {
                continue;
            }
            let note_count = count_md_files(&sub_path);
            notebooks.push(NotebookInfo {
                name: sub_name,
                path: sub_rel,
                stack: Some(name.clone()),
                note_count,
            });
        }
    }

    notebooks.sort_by(|a, b| a.path.cmp(&b.path));
    Ok(notebooks)
}
